### PR TITLE
Additional material outputs

### PR DESCRIPTION
**Pull request by jdannberg**
Add a structure that makes it possible to add new quantities to the MaterialModelOutputs.

### CODE REVIEW

File: include/aspect/material_model/interface.h [hunk 1]
```diff
@@ -365,6 +379,31 @@ namespace aspect
...
+       * Vector of shared pointers to additional material model output
+       * objects that can then be added to MaterialModelOutputs. By default,
+       * no outputs are added.
+       */
+      std::vector<std_cxx11::shared_ptr<AdditionalMaterialOutputs<dim> > > additional_outputs;
+
+      /**
+       * Given an additional material model output class, returns a pointer to this
+       * additional material model output object if it used in the current simulation.
+       * The output can then be filled in the evaluate() function. If the output does
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: Qualify the function name as `MaterialModels::Interface::evaluate()`.

File: include/aspect/material_model/interface.h [hunk 2]
```diff
@@ -365,6 +379,31 @@ namespace aspect
...
+
+      /**
+       * Vector of shared pointers to additional material model output
+       * objects that can then be added to MaterialModelOutputs. By default,
+       * no outputs are added.
+       */
+      std::vector<std_cxx11::shared_ptr<AdditionalMaterialOutputs<dim> > > additional_outputs;
+
+      /**
+       * Given an additional material model output class, returns a pointer to this
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: ...output class -> ...output class as explicitly specified template argument.

File: include/aspect/material_model/interface.h [hunk 3]
```diff
@@ -365,6 +379,31 @@ namespace aspect
...
+       * additional material model output object if it used in the current simulation.
+       * The output can then be filled in the evaluate() function. If the output does
+       * not exist, a null pointer is returned.
+       */
+      template <class T>
+      T *get_additional_output()
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: Would you mind renaming `T` to `AdditionalOutputType`?

File: include/aspect/material_model/interface.h [hunk 4]
```diff
@@ -365,6 +379,31 @@ namespace aspect
...
+      {
+        for (unsigned int i=0; i<additional_outputs.size(); ++i)
+          {
```
Comment (bangerth) [inline include/aspect/material_model/interface.h]: Please move the implementation of this function out of the class declaration and down to the bottom of the `.h` file.

**Comment by bangerth** [comment]
Nice testcase! I think I may not have bothered to add a testcase. It's good to know that you've become even more diligent than I usually am :-)

The patch looks good. Please make the few minor changes and merge -- thanks!

(@MFraters -- please take a look at the syntax here.)

**Comment by jdannberg** [comment]
Just waiting for the tester...
And the testcase was all Timo's doing.

**Pull request merged**